library/core/src/slice/ascii.rs

```diff
@@ -80,6 +80,82 @@ impl [u8] {
     pub fn escape_ascii(&self) -> EscapeAscii<'_> {
         EscapeAscii { inner: self.iter().flat_map(EscapeByte) }
     }
+
+    /// Returns a byte slice with leading ASCII whitespace bytes removed.
+    ///
+    /// 'Whitespace' refers to the definition used by
+    /// `u8::is_ascii_whitespace`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(byte_slice_trim_ascii)]
+    ///
+    /// assert_eq!(b" \t hello world\n".trim_start_ascii(), b"hello world\n");
+    /// assert_eq!(b"  ".trim_start_ascii(), b"");
+    /// assert_eq!(b"".trim_start_ascii(), b"");
+    /// ```
+    #[unstable(feature = "byte_slice_trim_ascii", issue = "none")]
+    pub fn trim_start_ascii(&self) -> &[u8] {
+        let mut i = 0;
+        while i < self.len() {
+            if self[i].is_ascii_whitespace() {
+                i += 1;
+            } else {
+                break;
+            }
+        }
+        // SAFETY: Index must be in the range 0..len
+        unsafe { self.get_unchecked(i..) }
+    }
```

Comment:
```suggestion
    pub const fn trim_start_ascii(&self) -> &[u8] {
        let mut rest = self;
        while let [c, bytes @ ..] = rest {
            if c.is_ascii_whitespace() {
                rest = bytes;
            } else {
                break;
            }
        }

        rest
    }
```

This appears to make the function const and almost for free. It does take a slightly different pattern in the generated assembly when compared in [Godbolt](https://rust.godbolt.org/z/oTbbsz3Kv), likely due to the `get_unchecked` rule, but this does not seem to be a real pessimization. The function just has two exit points instead of one, so the generated code is ever-so-slightly different. The path through the main function is the same, and the early exit seems actually shorter, in isolation, so I am not confident it doesn't improve the long tail of no-op cases. Something to try out on some benchmarks when const traits land, perhaps.

Reply:
That's a very interesting approach in the style of functional programming languages. I definitely see benefit from making the functions const, so I'd say it's worthwhile to change the implementation.

Reply:
Implementation has been updated with this approach (with some renamed variables to improve clarity). @workingjubilee I mentioned you in the "Co-authored-by" commit attribute.